### PR TITLE
ceph: added support for multus for csi

### LINKED ISSUE #5356: Multus and CSI
CSI templates need to see their annotation updated is Multus is detected as a provider.

### PR DESCRIPTION
**Pull request by rohan47**
**Description of your changes:**
CSI pods now utilize multus networking and connect to public
network specified in the CephCluster CR.

Signed-off-by: rohan47 <rohgupta@redhat.com>
**Which issue is resolved by this Pull Request:**
Resolves #5356 

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

[test ceph]

### CODE REVIEW

File: pkg/operator/ceph/csi/spec.go [hunk 1]
```diff
@@ -378,6 +379,10 @@ func startDrivers(namespace string, clientset kubernetes.Interface, ver *version
 		// apply resource request and limit to rbdplugin containers
 		applyResourcesToContainers(clientset, rbdPluginResource, &rbdPlugin.Spec.Template.Spec)
 		k8sutil.SetOwnerRef(&rbdPlugin.ObjectMeta, ownerRef)
+		err = applyCephClusterNetworkConfig(&rbdPlugin.Spec.Template.ObjectMeta, rookclientset)
```
Comment (rohan47) [inline pkg/operator/ceph/csi/spec.go:382]: The annotations are applied to the csi daemonset pods but the `k8s.v1.cni.cncf.io/networks-status` annotation is missing so couldn't verify if the pods are using multus

Comment (leseb) [reply pkg/operator/ceph/csi/spec.go:382]: Is the annotation getting overwritten?

File: pkg/operator/ceph/csi/spec.go [hunk 2]
```diff
@@ -406,6 +411,10 @@ func startDrivers(namespace string, clientset kubernetes.Interface, ver *version
 			Type: apps.RecreateDeploymentStrategyType,
 		}
 
+		err = applyCephClusterNetworkConfig(&rbdProvisionerDeployment.Spec.Template.ObjectMeta, rookclientset)
+		if err != nil {
+			return errors.Wrapf(err, "failed to apply network config to rbd plugin provisioner deployment: %+v", rbdPlugin)
```
Comment (leseb) [inline pkg/operator/ceph/csi/spec.go:416]:
```suggestion
			return errors.Wrapf(err, "failed to apply network config to rbd provisioner deployment: %+v", rbdProvisionerDeployment)
```

File: pkg/operator/ceph/csi/spec.go [hunk 3]
```diff
@@ -426,6 +435,10 @@ func startDrivers(namespace string, clientset kubernetes.Interface, ver *version
 		// apply resource request and limit to cephfs plugin containers
 		applyResourcesToContainers(clientset, cephFSPluginResource, &cephfsPlugin.Spec.Template.Spec)
 		k8sutil.SetOwnerRef(&cephfsPlugin.ObjectMeta, ownerRef)
+		err = applyCephClusterNetworkConfig(&cephfsPlugin.Spec.Template.ObjectMeta, rookclientset)
+		if err != nil {
+			return errors.Wrapf(err, "failed to apply network config to cephfs plugin daemonset: %+v", rbdPlugin)
```
Comment (leseb) [inline pkg/operator/ceph/csi/spec.go:440]:
```suggestion
			return errors.Wrapf(err, "failed to apply network config to cephfs plugin daemonset: %+v", cephfsPlugin)
```

File: pkg/operator/ceph/csi/spec.go [hunk 4]
```diff
@@ -455,6 +468,11 @@ func startDrivers(namespace string, clientset kubernetes.Interface, ver *version
 		cephfsProvisionerDeployment.Spec.Strategy = apps.DeploymentStrategy{
 			Type: apps.RecreateDeploymentStrategyType,
 		}
+
+		err = applyCephClusterNetworkConfig(&cephfsProvisionerDeployment.Spec.Template.ObjectMeta, rookclientset)
+		if err != nil {
+			return errors.Wrapf(err, "failed to apply network config to cephfs plugin provisioner deployment: %+v", rbdPlugin)
```
Comment (leseb) [inline pkg/operator/ceph/csi/spec.go:474]:
```suggestion
			return errors.Wrapf(err, "failed to apply network config to cephfs plugin provisioner deployment: %+v", cephfsProvisionerDeployment)
```

File: pkg/operator/ceph/csi/spec.go [hunk 5]
```diff
@@ -482,6 +500,20 @@ func startDrivers(namespace string, clientset kubernetes.Interface, ver *version
 	return nil
 }
 
+func applyCephClusterNetworkConfig(objectMeta *metav1.ObjectMeta, rookclientset rookclient.Interface) error {
+	cephClusters, err := rookclientset.CephV1().CephClusters(objectMeta.Namespace).List(metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
```
Comment (leseb) [inline pkg/operator/ceph/csi/spec.go:507]: return nil if there is no cluster and print a debug message?

File: pkg/operator/k8sutil/network.go [hunk 6]
```diff
@@ -119,7 +119,14 @@ func ApplyMultus(net rookv1.NetworkSpec, objectMeta *metav1.ObjectMeta) error {
 			shortSyntax = true
 		}
 
-		v = append(v, string(ns))
+		isCsi := strings.Contains(objectMeta.Labels["app"], "csi-")
+		if isCsi {
+			if k == "public" {
```
Comment (leseb) [inline pkg/operator/k8sutil/network.go:124]: I think we have a constant for this somewhere, don't we?

File: pkg/operator/ceph/csi/spec.go (hunk 1 above)
Comment (rohan47) [reply pkg/operator/ceph/csi/spec.go:382]: No.
Just that there is no `k8s.v1.cni.cncf.io/networks-status` annotation in the pod and I am not able to get in the containers to check if the multus interfaces are added

Comment (rohan47) [reply pkg/operator/ceph/csi/spec.go:382]: The annotations are applied normally

File: pkg/operator/ceph/csi/spec.go (hunk 2 above)
Comment (rohan47) [reply pkg/operator/ceph/csi/spec.go:416]: Ah, sorry, copy-paste mistake. Updating.

File: pkg/operator/ceph/csi/spec.go (hunk 1 above)
Comment (leseb) [reply pkg/operator/ceph/csi/spec.go:382]: Well how can the annotation be applied normally if not visible later on? Something is off.

Comment (rohan47) [reply pkg/operator/ceph/csi/spec.go:380]: The annotation is applied
```
annotations:
    k8s.v1.cni.cncf.io/networks: public
    openshift.io/scc: rook-ceph-csi
```

There should be a `k8s.v1.cni.cncf.io/networks-status`
something like the following:
```
Annotations:  k8s.v1.cni.cncf.io/networks: public
              k8s.v1.cni.cncf.io/networks-status:
                [{
                    "name": "openshift-sdn",
                    "interface": "eth0",
                    "ips": [
                        "10.128.2.31"
                    ],
                    "dns": {},
                    "default-route": [
                        "10.128.2.1"
                    ]
                },{
                    "name": "public",
                    "interface": "net1",
                    "ips": [
                        "192.168.1.209"
                    ],
                    "mac": "4e:5d:a8:b7:36:2b",
                    "dns": {}
                }]
              openshift.io/scc: rook-ceph-csi```

**Comment by Madhu-1** [comment]
I have 2 questions related to multus support.
1. As the multus is per ceph cluster and a CSI driver is deployed once per rook operator, a rook operator can manage multiple ceph clusters. what happens if different ceph cluster uses different multus network?
2. Ceph CSI daemonset pods run with host network as we are applying the multus to daemonset pods, how it's going to behave, is daemonset pods going to use multus instead of host network?

cc @travisn

**Comment by rohan47** [comment]
> I have 2 questions related to multus support.
> 
>     1. As the multus is per ceph cluster and a CSI driver is deployed once per rook operator, a rook operator can manage multiple ceph clusters. what happens if different ceph cluster uses different multus network?
>  
After thinking about this it looks like the annotatins will get replaced.
Will fix this.
>     2. Ceph CSI daemonset pods run with host network as we are applying the multus to daemonset pods, how it's going to behave, is daemonset pods going to use multus instead of host network?
> 
This is the part where I was stuck. Trying to use multus instead of host network and will update once I have the results.
> 
> cc @travisn

**Comment by rohan47** [comment]
@Madhu-1 csi daemonset doesn't work on multus network. What was the reason for using host network for the daemonsets?
What will be the drawback if we use a host network for daemonset pods and multus for the rest of the csi pods?

**Comment by travisn** [comment]
> With this PR we can remove the "experimental" flag from the doc https://rook.io/docs/rook/v1.3/ceph-cluster-crd.html and add a mention in the release note too.

Before we remove the experimental flag, how about we get 3rd party validation as suggested in the separate email thread today.

File: pkg/operator/k8sutil/network.go [hunk 7]
```diff
@@ -26,6 +26,13 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+const (
+	// PublicNetworkSelectorKeyName is the network selector key for the ceph public network
+	PublicNetworkSelectorKeyName = "public"
```
Comment (leseb) [inline pkg/operator/k8sutil/network.go:31]: This should not move here since the k8sutil package is generic for all the operators, not specific to Ceph.

File: pkg/operator/k8sutil/network.go [hunk 8]
```diff
@@ -34,7 +41,7 @@ type NetworkAttachmentConfig struct {
 	Mode       string `json:"mode"`
 	Ipam       struct {
 		Type       string `json:"type"`
-		Subnet     string `json:"subnet"`
+		Range      string `json:"range"`
```
Comment (leseb) [inline pkg/operator/k8sutil/network.go:44]: Make this a separate commit.

Comment (rohan47) [reply pkg/operator/k8sutil/network.go:44]: I think this should be in a separate PR and should be rebased to the releases that have multus support

Comment (rohan47) [reply pkg/operator/k8sutil/network.go:44]: https://github.com/rook/rook/pull/5864

**Comment by rohan47** [comment]
This doc has the connection information of mons and OSDs and the log of the failed OSD

**Comment by rohan47** [comment]
This doc has the connection information of mons and osds and the [logs](https://docs.google.com/document/d/1mn6Q6rpRKZNHWPzJNNP4fpDE5YNhPxoRkWF4jYGHGEQ/edit?usp=sharing) of the failed OSD

**Comment by rohan47** [comment]
> > With this PR we can remove the "experimental" flag from the doc https://rook.io/docs/rook/v1.3/ceph-cluster-crd.html and add a mention in the release note too.
> 
> Before we remove the experimental flag, how about we get 3rd party validation as suggested in the separate email thread today.

We are keeping the experimental flag right?

**Comment by leseb** [comment]
> > > With this PR we can remove the "experimental" flag from the doc https://rook.io/docs/rook/v1.3/ceph-cluster-crd.html and add a mention in the release note too.
> > 
> > 
> > Before we remove the experimental flag, how about we get 3rd party validation as suggested in the separate email thread today.
> 
> We are keeping the experimental flag right?

Yes and let's link the ceph-csi issue too.

**Comment by rohan47** [comment]
> > > > With this PR we can remove the "experimental" flag from the doc https://rook.io/docs/rook/v1.3/ceph-cluster-crd.html and add a mention in the release note too.
> > > 
> > > 
> > > Before we remove the experimental flag, how about we get 3rd party validation as suggested in the separate email thread today.
> > 
> > 
> > We are keeping the experimental flag right?
> 
> Yes and let's link the ceph-csi issue too.



> > > > With this PR we can remove the "experimental" flag from the doc https://rook.io/docs/rook/v1.3/ceph-cluster-crd.html and add a mention in the release note too.
> > > 
> > > 
> > > Before we remove the experimental flag, how about we get 3rd party validation as suggested in the separate email thread today.
> > 
> > 
> > We are keeping the experimental flag right?
> 
> Yes and let's link the ceph-csi issue too.

Okay

**Comment by leseb** [comment]
> > > > > With this PR we can remove the "experimental" flag from the doc https://rook.io/docs/rook/v1.3/ceph-cluster-crd.html and add a mention in the release note too.
> > > > 
> > > > 
> > > > Before we remove the experimental flag, how about we get 3rd party validation as suggested in the separate email thread today.
> > > 
> > > 
> > > We are keeping the experimental flag right?
> > 
> > 
> > Yes and let's link the ceph-csi issue too.
> 
> > > > > With this PR we can remove the "experimental" flag from the doc https://rook.io/docs/rook/v1.3/ceph-cluster-crd.html and add a mention in the release note too.
> > > > 
> > > > 
> > > > Before we remove the experimental flag, how about we get 3rd party validation as suggested in the separate email thread today.
> > > 
> > > 
> > > We are keeping the experimental flag right?
> > 
> > 
> > Yes and let's link the ceph-csi issue too.
> 
> Okay

Any updates @rohan47? We should move forward with that.

File: pkg/operator/ceph/csi/spec.go [hunk 9]
```diff
@@ -498,6 +522,24 @@ func deleteCSIDriverResources(
 	return succeeded
 }
 
+func applyCephClusterNetworkConfig(objectMeta *metav1.ObjectMeta, rookclientset rookclient.Interface) (bool, error) {
+	cephClusters, err := rookclientset.CephV1().CephClusters(objectMeta.Namespace).List(metav1.ListOptions{})
+	if err != nil {
+		logger.Debugf("No CephClusters found in namespace %s.", objectMeta.Namespace)
+		return false, nil
```
Comment (leseb) [inline pkg/operator/ceph/csi/spec.go:529]:
```suggestion
		return false, errors.New(...)
```

Return an error otherwise nothing will be printed in `startDrivers()`.

File: pkg/operator/ceph/csi/spec.go [hunk 10]
```diff
@@ -498,6 +522,24 @@ func deleteCSIDriverResources(
...
+	}
+	for _, cephCluster := range cephClusters.Items {
+		if cephCluster.Spec.Network.IsMultus() {
+			err = k8sutil.ApplyMultus(cephCluster.Spec.Network.NetworkSpec, objectMeta)
+			if err != nil {
+				return false, err
```
Comment (leseb) [inline pkg/operator/ceph/csi/spec.go:535]: wrap the error

**Review by leseb — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/csi/spec.go [hunk 11]
```diff
@@ -498,6 +522,23 @@ func deleteCSIDriverResources(
 	return succeeded
 }
 
+func applyCephClusterNetworkConfig(objectMeta *metav1.ObjectMeta, rookclientset rookclient.Interface) (bool, error) {
+	cephClusters, err := rookclientset.CephV1().CephClusters(objectMeta.Namespace).List(metav1.ListOptions{})
+	if err != nil {
+		return false, errors.New(fmt.Sprintf("failed to find CephClusters in namespace %q", objectMeta.Namespace))
```
Comment (leseb) [inline pkg/operator/ceph/csi/spec.go:528]: Use `errors.Errorf()` instead of `fmt.Sprintf()` inside `errors.New()`.

**Review by leseb — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/csi/spec.go (hunk 11 above)
Comment (rohan47) [reply pkg/operator/ceph/csi/spec.go:528]: `errors.Errorf()` returns error so it can' t be used inside `errors.New()`
Using `errors.Errorf()` instead of `errors.New()`

Comment (leseb) [reply pkg/operator/ceph/csi/spec.go]: Maybe I wasn't clear, just use `errors.Errorf()`

**Review by leseb — APPROVED** [review]

File: pkg/operator/ceph/csi/spec.go (hunk 11 above)
Comment (rohan47) [reply pkg/operator/ceph/csi/spec.go]: done

File: pkg/operator/ceph/csi/spec.go (hunk 9 above)
Comment (rohan47) [reply pkg/operator/ceph/csi/spec.go]: done

File: pkg/operator/ceph/csi/spec.go (hunk 10 above)
Comment (rohan47) [reply pkg/operator/ceph/csi/spec.go]: done

File: pkg/operator/ceph/csi/spec.go [hunk 12]
```diff
@@ -498,6 +522,23 @@ func deleteCSIDriverResources(
...
+	for _, cephCluster := range cephClusters.Items {
+		if cephCluster.Spec.Network.IsMultus() {
+			err = k8sutil.ApplyMultus(cephCluster.Spec.Network.NetworkSpec, objectMeta)
+			if err != nil {
+				return false, errors.Wrapf(err, "failed to apply multus configuration to CephCluster %q", cephCluster.Name)
+			}
+		}
+	}
+
+	return true, nil
```
Comment (Madhu-1) [inline pkg/operator/ceph/csi/spec.go:539]: even if the ceph cluster is not using multus it's returning true? 
don't we need to return false if the ceph cluster network is not multus?

File: pkg/operator/ceph/csi/spec.go [hunk 13]
```diff
@@ -498,6 +522,23 @@ func deleteCSIDriverResources(
 	return succeeded
 }
 
+func applyCephClusterNetworkConfig(objectMeta *metav1.ObjectMeta, rookclientset rookclient.Interface) (bool, error) {
+	cephClusters, err := rookclientset.CephV1().CephClusters(objectMeta.Namespace).List(metav1.ListOptions{})
```
Comment (Madhu-1) [inline pkg/operator/ceph/csi/spec.go:526]: question: won't this fail if the ceph cluster and CSI is in a different namespace?

Comment (rohan47) [reply pkg/operator/ceph/csi/spec.go:526]: yes it will fail

File: pkg/operator/ceph/csi/spec.go (hunk 12 above)
Comment (leseb) [reply pkg/operator/ceph/csi/spec.go:539]: Good point, we should only return if multus got configured, I'm sending a PR to fix this.

Comment (leseb) [reply pkg/operator/ceph/csi/spec.go:539]: https://github.com/rook/rook/pull/6152

File: pkg/operator/ceph/csi/spec.go (hunk 13 above)
Comment (rohan47) [reply pkg/operator/ceph/csi/spec.go:526]: Yes. 
The namespace we have here is actually the Rook Ceph operators namespace, right?

Comment (Madhu-1) [reply pkg/operator/ceph/csi/spec.go:526]: Yes its rook ceph operator namespace
